DATA INGESTION

In [1]:
from datasets import load_dataset

dataset = load_dataset("go_emotions")
print(dataset)

c:\UDEMY COURSE-ML,DL,NLP\Youtube_Comments_Sentiment_Analysis\venv\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


DatasetDict({
    train: Dataset({
        features: ['text', 'labels', 'id'],
        num_rows: 43410
    })
    validation: Dataset({
        features: ['text', 'labels', 'id'],
        num_rows: 5426
    })
    test: Dataset({
        features: ['text', 'labels', 'id'],
        num_rows: 5427
    })
})


In [2]:
dataset.shape

{'train': (43410, 3), 'validation': (5426, 3), 'test': (5427, 3)}

In [3]:
train_dataset=dataset["train"]
test_dataset=dataset["test"]
val_dataset=dataset["validation"]

In [4]:
print(train_dataset[0])
print(val_dataset[0])
print(test_dataset[0])


{'text': "My favourite food is anything I didn't have to cook myself.", 'labels': [27], 'id': 'eebbqej'}
{'text': 'Is this in New Orleans?? I really feel like this is New Orleans.', 'labels': [27], 'id': 'edgurhb'}
{'text': 'I’m really sorry about your situation :( Although I love the names Sapphira, Cirilla, and Scarlett!', 'labels': [25], 'id': 'eecwqtt'}


DATA PREPROCESSING

In [5]:
label_names = dataset["train"].features["labels"].feature.names

label_names

['admiration',
 'amusement',
 'anger',
 'annoyance',
 'approval',
 'caring',
 'confusion',
 'curiosity',
 'desire',
 'disappointment',
 'disapproval',
 'disgust',
 'embarrassment',
 'excitement',
 'fear',
 'gratitude',
 'grief',
 'joy',
 'love',
 'nervousness',
 'optimism',
 'pride',
 'realization',
 'relief',
 'remorse',
 'sadness',
 'surprise',
 'neutral']

In [6]:
def decode_labels(example):
    example["emotions"]=[label_names[i] for i in example["labels"]]
    return example

In [7]:
train_ds = dataset["train"].map(decode_labels)
val_ds   = dataset["validation"].map(decode_labels)
test_ds  = dataset["test"].map(decode_labels)

In [8]:
train_ds[0]

{'text': "My favourite food is anything I didn't have to cook myself.",
 'labels': [27],
 'id': 'eebbqej',
 'emotions': ['neutral']}

In [9]:
train_ds = train_ds.remove_columns(["id", "labels"])
val_ds   = val_ds.remove_columns(["id", "labels"])
test_ds  = test_ds.remove_columns(["id", "labels"])

In [10]:
print(train_ds[0])
print(test_ds[0])
print(val_ds[0])

{'text': "My favourite food is anything I didn't have to cook myself.", 'emotions': ['neutral']}
{'text': 'I’m really sorry about your situation :( Although I love the names Sapphira, Cirilla, and Scarlett!', 'emotions': ['sadness']}
{'text': 'Is this in New Orleans?? I really feel like this is New Orleans.', 'emotions': ['neutral']}


In [11]:
from datasets import Dataset

train_ds = train_ds.map(lambda x: {'text': x['text'].lower()})
test_ds = test_ds.map(lambda x: {'text': x['text'].lower()})
val_ds = val_ds.map(lambda x: {'text': x['text'].lower()})

In [12]:
train_ds[0]

{'text': "my favourite food is anything i didn't have to cook myself.",
 'emotions': ['neutral']}

In [13]:
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences

tokenizer = Tokenizer(num_words=20000, oov_token="<OOV>")

tokenizer.fit_on_texts(train_ds['text'])

c:\UDEMY COURSE-ML,DL,NLP\Youtube_Comments_Sentiment_Analysis\venv\lib\site-packages\google\api_core\_python_version_support.py:275: FutureWarning: You are using a Python version (3.10.0) which Google will stop supporting in new releases of google.api_core once it reaches its end of life (2026-10-04). Please upgrade to the latest Python version, or at least Python 3.11, to continue receiving updates for google.api_core past that date.
  warnings.warn(message, FutureWarning)


In [14]:
len(tokenizer.word_index)

28208

In [15]:
tokenizer.word_index

{'<OOV>': 1,
 'the': 2,
 'i': 3,
 'to': 4,
 'a': 5,
 'you': 6,
 'and': 7,
 'is': 8,
 'that': 9,
 'name': 10,
 'it': 11,
 'of': 12,
 'this': 13,
 'in': 14,
 'for': 15,
 'was': 16,
 'my': 17,
 'not': 18,
 'but': 19,
 'be': 20,
 'on': 21,
 'so': 22,
 'just': 23,
 'have': 24,
 'are': 25,
 'like': 26,
 'with': 27,
 'me': 28,
 'they': 29,
 'your': 30,
 'he': 31,
 'what': 32,
 'if': 33,
 'as': 34,
 'at': 35,
 "it's": 36,
 'no': 37,
 'all': 38,
 'about': 39,
 'do': 40,
 'love': 41,
 "i'm": 42,
 'one': 43,
 'get': 44,
 'people': 45,
 'we': 46,
 'good': 47,
 'would': 48,
 'can': 49,
 'how': 50,
 'out': 51,
 'up': 52,
 'from': 53,
 'an': 54,
 'really': 55,
 "don't": 56,
 'think': 57,
 'or': 58,
 'know': 59,
 'she': 60,
 'when': 61,
 'too': 62,
 'more': 63,
 'there': 64,
 'her': 65,
 'them': 66,
 'thanks': 67,
 'him': 68,
 'now': 69,
 'his': 70,
 'will': 71,
 'lol': 72,
 'time': 73,
 'has': 74,
 'some': 75,
 'see': 76,
 "that's": 77,
 'because': 78,
 'thank': 79,
 'who': 80,
 'much': 81,
 'why': 8

In [16]:
def tokenize_example(example):
    example['tokens'] = tokenizer.texts_to_sequences([example['text']])[0]
    return example

train_ds = train_ds.map(tokenize_example)
val_ds   = val_ds.map(tokenize_example)
test_ds  = test_ds.map(tokenize_example)

In [17]:
train_ds[0]

{'text': "my favourite food is anything i didn't have to cook myself.",
 'emotions': ['neutral'],
 'tokens': [17, 1201, 513, 8, 205, 3, 150, 24, 4, 3867, 362]}

In [18]:
lengths = [len(tokens) for tokens in train_ds['tokens']]

max_len = max(lengths)
max_len

142

In [19]:
import numpy as np

np.percentile(lengths, [90, 95, 99])

array([23., 24., 27.])

90% of sentences have <=23tokens
95% of sentences have <=24tokens
99% of sentences have <=27tokens

so lets take max_len of sentence=27
only 1 sentence has 142 length

In [20]:
max_len=27

In [21]:
X_train=pad_sequences(
    train_ds['tokens'],
    maxlen=max_len,
    padding='post'
)

X_val=pad_sequences(
    val_ds['tokens'],
    maxlen=max_len,
    padding='post'
)

X_test=pad_sequences(
    test_ds['tokens'],
    maxlen=max_len,
    padding='post'
)

In [22]:
X_train.shape

(43410, 27)

In [23]:
train_ds.features

{'text': Value('string'),
 'emotions': List(Value('string')),
 'tokens': List(Value('int64'))}

In [24]:
all_emotions = set()
for emotions in train_ds['emotions']:
    for e in emotions:
        all_emotions.add(e)

all_emotions = sorted(list(all_emotions))

num_labels = len(all_emotions)   

emotion2idx = {emotion: idx for idx, emotion in enumerate(all_emotions)}
idx2emotion = {idx: emotion for emotion, idx in emotion2idx.items()}


In [25]:
import numpy as np

def encode_multi_hot(example):
    label_vector = np.zeros(num_labels, dtype=np.float32)
    for emotion in example['emotions']:
        label_vector[emotion2idx[emotion]] = 1.0
    return {'label_vector': label_vector}


In [26]:
train_ds = train_ds.map(encode_multi_hot)
val_ds   = val_ds.map(encode_multi_hot)
test_ds  = test_ds.map(encode_multi_hot)

In [27]:
train_ds[40]

{'text': 'aw, thanks! i appreciate that! ',
 'emotions': ['admiration', 'gratitude'],
 'tokens': [1738, 67, 3, 418, 9],
 'label_vector': [1.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  1.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0]}

In [28]:
y_train = np.array(train_ds['label_vector'])
y_val   = np.array(val_ds['label_vector'])
y_test  = np.array(test_ds['label_vector'])

In [29]:
y_train[0]

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0.])

In [30]:
print(X_train.shape)
y_train.shape

(43410, 27)


(43410, 28)

MODEL TRAINING

In [31]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Dense, Dropout, Bidirectional, Input, SpatialDropout1D
from tensorflow.keras.regularizers import l2
from tensorflow.keras.callbacks import EarlyStopping, ReduceLROnPlateau

vocab_size = 28082
embed_dim = 100
max_len = 27

model = Sequential()
model.add(Input(shape=(max_len,)))

# 1. Embedding Layer
model.add(Embedding(vocab_size, embed_dim))

# 2. SpatialDropout (Better for NLP than standard Dropout)
# Drops entire 1D feature maps instead of individual elements
model.add(SpatialDropout1D(0.3))

# 3. First LSTM Layer
# Reduced units from 128 to 64 to prevent overfitting
model.add(
    Bidirectional(
        LSTM(64, return_sequences=True) 
    )
)

# 4. Second LSTM Layer
# Reduced units from 64 to 32
model.add(
    Bidirectional(
        LSTM(32)
    )
)

# 5. Stronger Dropout
model.add(Dropout(0.5))

# 6. Dense Layer with L2 Regularization
# kernel_regularizer=l2(0.001) penalizes large weights
model.add(Dense(64, activation='relu', kernel_regularizer=l2(0.001)))

# 7. Output Dropout
model.add(Dropout(0.5))

# 8. Output Layer
model.add(Dense(num_labels, activation='sigmoid'))

model.compile(
    loss='binary_crossentropy',
    optimizer='adam',
    metrics=['accuracy']
)

model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding (Embedding)           │ (None, 27, 100)        │     2,808,200 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ spatial_dropout1d               │ (None, 27, 100)        │             0 │
│ (SpatialDropout1D)              │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ bidirectional (Bidirectional)   │ (None, 27, 128)        │        84,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ bidirectional_1 (Bidirectional) │ (None, 64)             │        41,216 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 64)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 64)             │         4,160 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 64)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 28)             │         1,820 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,939,876 (11.21 MB)

 Trainable params: 2,939,876 (11.21 MB)

 Non-trainable params: 0 (0.00 B)

In [32]:
# EarlyStopping: Stops if no improvement for 5 epochs
early_stopping = EarlyStopping(
    monitor='val_loss',
    patience=5,
    restore_best_weights=True,
    verbose=1
)

# --- Training ---
history = model.fit(
    X_train,
    y_train,
    validation_data=(X_val, y_val),
    epochs=20,          # Increased slightly since we have EarlyStopping
    batch_size=64,
    callbacks=[early_stopping]
)

Epoch 1/20
679/679 ━━━━━━━━━━━━━━━━━━━━ 71s 91ms/step - accuracy: 0.2607 - loss: 0.2023 - val_accuracy: 0.2960 - val_loss: 0.1472
Epoch 2/20
679/679 ━━━━━━━━━━━━━━━━━━━━ 66s 97ms/step - accuracy: 0.3407 - loss: 0.1469 - val_accuracy: 0.3879 - val_loss: 0.1334
Epoch 3/20
679/679 ━━━━━━━━━━━━━━━━━━━━ 60s 88ms/step - accuracy: 0.3758 - loss: 0.1371 - val_accuracy: 0.3920 - val_loss: 0.1288
Epoch 4/20
679/679 ━━━━━━━━━━━━━━━━━━━━ 59s 87ms/step - accuracy: 0.3917 - loss: 0.1302 - val_accuracy: 0.4018 - val_loss: 0.1260
Epoch 5/20
679/679 ━━━━━━━━━━━━━━━━━━━━ 42s 62ms/step - accuracy: 0.4057 - loss: 0.1251 - val_accuracy: 0.3988 - val_loss: 0.1243
Epoch 6/20
679/679 ━━━━━━━━━━━━━━━━━━━━ 51s 75ms/step - accuracy: 0.4157 - loss: 0.1210 - val_accuracy: 0.3994 - val_loss: 0.1251
Epoch 7/20
679/679 ━━━━━━━━━━━━━━━━━━━━ 62s 91ms/step - accuracy: 0.4287 - loss: 0.1175 - val_accuracy: 0.3970 - val_loss: 0.1274
Epoch 8/20
679/679 ━━━━━━━━━━━━━━━━━━━━ 60s 88ms/step - accuracy: 0.4405 - loss: 0.1142 - 

In [33]:
model.evaluate(X_test, y_test)

170/170 ━━━━━━━━━━━━━━━━━━━━ 4s 22ms/step - accuracy: 0.4022 - loss: 0.1236


[0.12362935394048691, 0.4022480249404907]

In [34]:
y_pred = model.predict(X_test)

170/170 ━━━━━━━━━━━━━━━━━━━━ 8s 31ms/step


In [35]:
y_pred

array([[6.30478710e-02, 1.97520759e-02, 1.63862091e-02, ...,
        8.92075971e-02, 1.12537533e-01, 6.45272508e-02],
       [1.81990951e-01, 8.80564600e-02, 1.55894337e-02, ...,
        2.03012861e-02, 4.49402221e-02, 5.27390018e-02],
       [1.34920493e-01, 5.60842007e-02, 1.01491790e-02, ...,
        5.64374886e-02, 6.99993670e-02, 5.64255528e-02],
       ...,
       [9.61924158e-03, 1.94181968e-03, 4.56755124e-02, ...,
        5.90747339e-04, 9.10792872e-03, 9.15566087e-03],
       [2.30609983e-01, 1.20598696e-01, 4.59844014e-03, ...,
        2.46913210e-02, 3.39878015e-02, 3.90134752e-02],
       [1.48650110e-02, 3.34663573e-03, 2.29960904e-02, ...,
        1.90829742e-04, 3.99234286e-03, 5.69447083e-03]],
      shape=(5427, 28), dtype=float32)

In [36]:
threshold = 0.5

def decode_prediction(prob_vector):
    emotions = []
    for idx, prob in enumerate(prob_vector):
        if prob >= threshold:
            emotions.append(idx2emotion[idx])
    return emotions


In [37]:
sample_id = 22

probs = y_pred[sample_id]
print(probs)
predicted_emotions = decode_prediction(probs)

print(predicted_emotions)

[9.8814415e-03 1.9821208e-03 2.1477319e-02 4.7309838e-02 6.3275620e-02
 1.0905957e-02 3.5502333e-02 6.5715194e-02 2.1396074e-03 1.0782666e-02
 5.6162223e-02 3.9409362e-03 1.0039188e-03 3.5375482e-03 1.2967985e-03
 2.9551864e-04 1.7462524e-04 8.0488244e-04 1.2970887e-03 2.8718598e-04
 7.0625436e-01 3.4588738e-03 1.7579335e-04 2.0152146e-02 2.7720883e-04
 1.1997963e-04 3.0989586e-03 4.1834451e-03]
['neutral']


In [38]:
def decode_text(token_ids):
    words = []
    for token in token_ids:
        if token != 0:  
            words.append(tokenizer.index_word.get(token, "<UNK>"))
    return " ".join(words)

In [39]:
i = 12
def decode_prediction_topk(prob_vector, k=2):
    top_indices = prob_vector.argsort()[-k:][::-1]
    return [idx2emotion[i] for i in top_indices]

# Original text (decoded from tokens)
text = decode_text(X_test[i])

# Model prediction
probs = y_pred[i]
emotions = decode_prediction_topk(probs)

print("Comment:")
print(text)

print("\nPredicted emotions:")
print(emotions)

Comment:
translation i wish i could afford it

Predicted emotions:
['optimism', 'sadness']


In [40]:
X_test[69]

array([  1,  73,   8, 293,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0], dtype=int32)

In [41]:
import pickle

# 1. Save the Keras Model
model.save('go_emotions_model.h5')
print("Model saved as 'go_emotions_model.h5'")

Model saved as 'go_emotions_model.h5'


In [42]:
with open('tokenizer.pickle', 'wb') as handle:
    pickle.dump(tokenizer, handle, protocol=pickle.HIGHEST_PROTOCOL)
print("Tokenizer saved as 'tokenizer.pickle'")

Tokenizer saved as 'tokenizer.pickle'
